Bugfix/issue11 generating 2024.1 models

AmateurECE opened the pull request:
This PR will resolve #11. Model generation succeeds, but the generator is currently mis-typing struct properties in the generated code. This will be resolved by adding a new ModelTypeMapper specialization in the generator. I will remove the "Draft" status from the PR when everything is fixed, and merge.

AmateurECE committed:
build: Update to 2024.2 models

Update the build script to pull the 2024.2 models, and update the
patches to be compatible. Code generation is failing at the moment,
caused by an exception in the code generator.

AmateurECE committed:
generator: Refactor ModelResolver to use IModelModelMapper

This was an existing point of abstraction created earlier and expressly
for this purpose, but I guess I just forgot to perform the refactor
earlier. This interface facilitates the future creation of new
transformations on the model graph that map one existing model
transparently to another, and can be used, e.g., to "erase" models from
the graph.

pmundt commented on generate.mk:
Note that 2024.2 has since been released. The Swordfish release should also be updated for the latest, which is presently v1.2.6.

AmateurECE replied:
Thank you! I've updated the Swordfish and Redfish models to these versions, though the registries (DSP8011) are still the latest at 2024.1 for the moment.

AmateurECE commented:
Sorry for the delay. I think I've _almost_ got this figured out. Turns out that a lot of plumbing types became inline schemas between 2023.1 and 2024.1, where previously there were only two OpenAPI inline schemas in the whole spec. That wasn't something we were set up to cleanly accommodate, so it required a little bit of work. I should be done soon!

AmateurECE committed:
generator: Use repository for reverse model resolution

Previously, model-type mappers were required to support a reverse
operation--a query that would produce the schema model name given the
Rust type. I had been putting off this refactor for a while, but now
that there's a need to map inline schema names to types (by definition a
lossy function), this constraint is no longer possible to enforce for
all model-type mappers. Instead, implement a repository "cache" in the
ModelResolver to allow these two to be associated the first time they're
resolved in the forward direction.

AmateurECE committed:
generator: Remove DocumentParser

This code, which was borrowed from swagger-codegen, flattens inline
models to become first-class schemas. It was very useful when there were
a number of inline schemas that had properties and required additional
struct definitions. In 2024.2, however, there are many inline structs
that can be rendered simply using std::option::Option, and so this code
has become more of a hindrance.

AmateurECE committed:
generator: Fix regex typo in UnversionedSchemaIdentifier

This sneaky character class typo lived in the codebase for 17 months
before it created a failure loud enough that I realized.

AmateurECE committed:
generator: Fix resource::Health query path

Previously, the ModelResolver would accept a full path for a model name
to resolve to a Rust type. This was, however, accidental, and recent
changes made this no longer possible. Change to using the model name
only.

AmateurECE committed:
generator: Add a number of new special abbreviations

These abbreviations only appeared in the Redfish data model as of
2024.1+.

AmateurECE committed:
generator: Override toString in identifiers

AmateurECE committed:
generator: Fix warning about unused imports

AmateurECE committed:
generator: Unit structs should derive Default

This is needed to satisfy the definition of Default for some
automatic derivations on other structs that include these as attributes.

AmateurECE committed:
generator: Refactor ModelResolver::getSchemaIdentifier

Previously this method took a Schema and explicitly extracted the $ref.
Now, however, I have a use case to use it where a Schema is not
available, so refactor it to take the $ref instead.

AmateurECE committed:
generator: Promoted inline schemas

This logic replaces the old DocumentParser class. Inline schemas that
represent algebraic product types, e.g. structs with members, need to be
promoted to "first-class" schemas and given a definition and schema ref,
so that their definitions can be generated. This commit introduces the
logic to do this in OpenapiSpecification, and a model-type mapper that
resolves the created $refs to types.

AmateurECE committed:
generator: Resolve some inline schemas to std types

There are a number of inline schemas in the Redfish data model that
represent simple types--for example, there's a repeated pattern that
specifies a property may either be of some type (where a $ref is
provided) or "null". Obviously, we already have a concept for that
(Option<>), so create a resolver that can recognize these primitive
patterns.

AmateurECE committed:
generator: HACK for swagger-parser#1961

Since I'm committed at this point to a Rust rewrite, implement a hack to
allow model generation to proceed without requiring a patched version of
swagger-parser.

AmateurECE committed:
models: Update to v2024.2

Update all of the infrastructure and examples to use the v2024.2 models.

AmateurECE committed:
generator: Don't abort if there's no JSON schema file

Previously, we would throw a RuntimeException if we couldn't identify
the JSON Schema file for a particular model. This is because in the
general case, services must indicate the JSON schema file for a model in
the headers of a response. However, some models (e.g.
odata_v4::ServiceDocument) actually aren't represented in the JSON
schema. Now, we allow these models to be generated without
implementations for Metadata<'static>.

AmateurECE committed:
generator: Fix uncaught CaseConversionError

This would occur while trying to identify the JSON schema file that
describes a particular model. Some models simply aren't described in
a JSON schema file.

AmateurECE committed:
generator: Fix NullPointerException in PascalCaseName

Previously, if we attempted to create a PascalCaseName with an empty ""
identifier, we would see a NullPointerException. Now, we see a
CaseConversionError instead.

AmateurECE commented:
Alright, @pmundt, I believe this PR is ready to be used! The `seuss` crate currently doesn't compile and I will want to fix that quick before merging (which I will do tonight), but I don't imagine that will block you for the time being. I also implemented a workaround in 9d5e512 that will allow us to merge #14 after this one is merged.

Please let me know if you experience any problems with it!

AmateurECE committed:
generator: Allow to disable property deletion in ModelDeletionPolicy

The ModelDeletionPolicy was designed to erase models from the
collection--this includes properties of other schemas that reference the
models. In some cases, however, we don't want this. Introduce a property
that allows to disable this behavior, so that only the models themselves
are deleted.

pmundt commented:
Just to follow up on this, everything looks good, and I'm able to build on the new schemas without any problems. Thanks for your efforts here!
